org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ImportRootProjectOperation.java

```diff
@@ -0,0 +1,122 @@
+        this.newProjectHandler = Preconditions.checkNotNull(newProjectHandler);
+    }
+
+    public void run(IProgressMonitor monitor) {
+        try {
+            runInWorkspace(monitor);
+        } catch (Exception e) {
+            throw new ImportRootProjectException(e);
+        }
+    }
+
+    private void runInWorkspace(IProgressMonitor monitor) throws CoreException {
+        ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
+
+            @Override
+            public void run(IProgressMonitor monitor) throws CoreException {
+                SubMonitor progress = SubMonitor.convert(monitor);
+                progress.setTaskName("Importing root project");
+                progress.setWorkRemaining(3);
+
+                File rootDir = ImportRootProjectOperation.this.buildConfiguration.getRootProjectDirectory();
+                verifyNoWorkspaceRootIsImported(rootDir, progress.newChild(1));
+                saveProjectConfiguration(ImportRootProjectOperation.this.buildConfiguration, rootDir, progress.newChild(1));
+                importRootProject(rootDir, progress.newChild(1));
+            }
+        }, monitor);
+    }
+
+    private void verifyNoWorkspaceRootIsImported(File rootDir, IProgressMonitor monitor) {
+        File workspaceRoot = ResourcesPlugin.getWorkspace().getRoot().getLocation().toFile();
+        if (rootDir.equals(workspaceRoot)) {
+            throw new UnsupportedConfigurationException(String.format("Project %s location matches workspace root %s", rootDir.getName(), workspaceRoot.getAbsolutePath()));
+        }
+    }
+
+    private void saveProjectConfiguration(BuildConfiguration buildConfiguration, File rootDir, IProgressMonitor monitor) {
+        ProjectConfiguration projectConfiguration = CorePlugin.configurationManager().createProjectConfiguration(buildConfiguration, rootDir);
+        CorePlugin.configurationManager().saveProjectConfiguration(projectConfiguration);
+    }
+
+    private void importRootProject(File rootDir, IProgressMonitor monitor) throws CoreException {
+        SubMonitor progress = SubMonitor.convert(monitor);
+        progress.setWorkRemaining(3);
+
+        WorkspaceOperations workspaceOperations = CorePlugin.workspaceOperations();
+        Optional<IProject> projectOrNull = workspaceOperations.findProjectByLocation(rootDir);
+
+        if (!projectOrNull.isPresent() && this.newProjectHandler.shouldImport()) {
+            String projectName = findFreeProjectName(rootDir.getName());
+            projectOrNull = Optional.of(workspaceOperations.createProject(projectName, rootDir, ImmutableList.<String>of(), progress.newChild(1)));
+        }
+
+        if (projectOrNull.isPresent()) {
+            IProject project = projectOrNull.get();
+            project.refreshLocal(IResource.DEPTH_INFINITE, progress.newChild(1));
+            CorePlugin.workspaceOperations().addNature(project, GradleProjectNature.ID, progress.newChild(1));
```

Comment:
❌ Why are we refreshing and adding the Gradle nature here? I think this should be done by the next step (synchronization).

Reply:
Same as above.